docs/CONFIG.md

# Config Options

The following config options are build-time config options. 

Note that config options tagged with "🔔" may affect the weight tensors and/or their layouts.


## Shape Config

Every entry within the "Shape Config" must be a three-integer tuple specifying the M/N/K dimensions.

- `problem_shape`:
  - In the problem shape, the M-dimension is consistently ignored because the actual value of M is dynamically determined based on the kernel input, rather than being set as a compile-time parameter.
  - The N and K dimensions represent the N and K of the weight matrix. In the case of MoE, these represent the N and K dimensions of an single expert's weight matrix.
  - `problem_shape.n` and `problem_shape.k` must be divisable by `block_shape.n` and `block_shape.k`, respectfully.

- `block_shape`: 
  - `m`: 
    - Must be divisable by 16.
    - `block_shape.m // warp_shape.m` must be powers of 2.
  - `n`:
    - Must be powers of 2.
    - Must be divisable by 64.
    - If use WGMMA, `block_shape.n // warp_shape.n` must be divisable by 4.
  - `k`:
    - Must be powers of 2.
    - Must be divisable by 32/64/128 for 16/8/4 bit activation, respectfully.

- `warp_shape`:
  - `m`: 
    - Must be divisable by 16.
    - `block_shape.m // warp_shape.m` must be powers of 2.
  - `n`:
    - Must be powers of 2.
    - 16 bit activation: must be 64. 8 bit activation: must be 64/32. 4 bit activation: must be 64/32/16.
    - If use WGMMA, `block_shape.n // warp_shape.n` must be divisable by 4.
  - `k`:
    - Must be powers of 2.
    - Must be divisable by 32/64/128 for 16/8/4 bit activation, respectfully.


## DataType Config 🔔

See support matrix in `README.md`.



## Quant Param Config


- `input_scale_group_size`: 
  - `0`: channelwise scale (also known as token-wise scale)
  - `>0`: must be powers of 2. The minimal value is 8/16/32 for 16/8/4-bit activation + MMA, and 16/32/64 for 16/8/4-bit activation + WGMMA.

- `weight_scale_group_size`🔔: 
  - `0`: channelwise scale
  - `>0`: must be powers of 2. The minimal value is 8/16/32 for 16/8/4-bit activation + MMA, and 16/32/64 for 16/8/4-bit activation + WGMMA.

- `has_zero_point`🔔: Enable dynamic zero point or not.
  - Must be `False` if `has_weight_scale` is False.
  - Share the same group size with `weight_scale_group_size`.


## Scheduler Config

- `use_stream_k`: Enable stream-k or not. It is recommended to be enabled in most case. If you want batch invariance mode, disable it.


## Pipeline Config

- `num_stages`: Num stages for multi-stages pipeline. Higher value requires larger shared memory.

- `use_warp_spec`: Enable warp specialization or not. Requires SM90+.

- `use_mbarrier`: Enable mbarrier for pipeline or not. Must be enabled for warp spec and/or tma. Can also be enabled if `use_cp_async` is True. Requires SM80+.

- `use_cp_async`: Enable `cp.async` or not. Requires SM80+.

- `use_tma`: Enable TMA or not. Requires SM90+.
  - `use_tma_a`, `use_tma_b`, `use_tma_c`, `use_tma_bs`, `use_tma_bzp`, `use_tma_bias`: You can also configure TMA for only a subset of tensors. Note that if any specific tensor is set to use TMA, the global `use_tma` flag must also be `True`. When `use_tma` is True, all options will default to `True` unless explicitly set to False or if they conflict with other configs.

## Epilogue Config

- `has_bias`: Enable bias fusion or not.

## MoE Config

- `is_moe`🔔: 
  - `True`: MoE GEMM. Note that humming currently only support triton-style indexed gemm, cublas-style grouped gemm is not ready now.
  - `False`: Dense GEMM.

- `top_k`: Must greater than 0 if `is_moe` is `True`.

- `is_moe_down`: The behavior of `is_moe_down`:

| `is_moe_down` | `True` | `False` |
| - | - | - |
| Read Index | `sorted_token_ids` | `sorted_token_ids // top_k` |
| Write Index | `sorted_token_ids` | `sorted_token_ids` |
| Multiply Topk Weights | YES | NO |



## MMA Config

- `mma_type`🔔: `mma` or `wgmma`
- `use_f16_accum`: Use float16 accumulation or not. This can have better performance since it requires less registers and mma with f16 accumulation have double tflops than mma with f32 accumulation on some devices (SM75/SM89). But it would increase the risk of numerical overflow. The f16 accum requires:
  - `c_dtype` must be `float16` (note that `bfloat16` is not supported)
  - `a_dtype` must be `float16` or `float8e4m3`. If `float8e4m3`, `b_dtype` must be integer type or floating type with exponent bits <= 3


humming/config/config.py
import dataclasses
import math
from typing import ClassVar

from typing_extensions import Self

from humming.config.base import BaseHummingConfig
from humming.config.enum import MmaType, WeightScaleType


@dataclasses.dataclass
class SchedulerConfig(BaseHummingConfig):
    use_stream_k: bool = True


@dataclasses.dataclass
class PipelineConfig(BaseHummingConfig):
    use_warp_spec: bool | None = None
    num_stages: int = 2
    num_ctas_per_sm: int = 1
    use_mbarrier: bool | None = None
    use_cp_async: bool | None = None
    use_tma: bool | None = None
    use_tma_a: bool | None = None
    use_tma_b: bool | None = None
    use_tma_c: bool | None = None
    use_tma_bs: bool | None = None
    use_tma_bzp: bool | None = None
    use_tma_bias: bool | None = None
    num_write_splits: int = 1
    multi_cast_size_a: int = 1
    multi_cast_size_b: int = 1

    _cpp_extra_names: ClassVar[tuple[str, ...]] = (
        "num_threads",
        "num_math_threads",
        "num_load_threads",
    )

    _name_map = {
        "use_mbarrier": "kUseMBarrier",
        "use_tma_bs": "kUseTmaBS",
        "use_tma_bzp": "kUseTmaBZP",
    }

    def __post_init__(self):
        self.num_threads = 0
        self.num_load_threads = 0
        self.num_math_threads = 0

    @classmethod
    def from_dict(cls, raw_config) -> Self:
        clean_config = cls._preprocess_dict(raw_config)
        config = cls(**clean_config)

        block_shape = raw_config["block_shape"]
        warp_shape = raw_config["warp_shape"]
        config.num_math_threads = math.prod(block_shape) // math.prod(warp_shape) * 32

        if config.use_warp_spec is None:
            config.use_warp_spec = False

        if config.use_warp_spec:
            config.num_load_threads = 128
            config.num_threads = config.num_math_threads + 128
        else:
            config.num_load_threads = config.num_math_threads
            config.num_threads = config.num_math_threads

        if config.use_tma is None:
            config.use_tma = False

        for name in dir(config):
            if not name.startswith("use_tma_"):
                continue
            if not config.use_tma:
                assert getattr(config, name) is not True
            if getattr(config, name) is None:
                setattr(config, name, config.use_tma)

        if config.use_mbarrier is None:
            config.use_mbarrier = False

        if config.use_cp_async is None:
            config.use_cp_async = raw_config["sm_version"] >= 80

        return config


@dataclasses.dataclass
class QuantParamConfig(BaseHummingConfig):
    input_scale_group_size: int = 0
    weight_scale_group_size: int = 0
    weight_scale_group_size_n: int = 0
    weight_scale_type: WeightScaleType | None = None
    use_int_weight_scale: bool = False
    has_zero_point: bool = False
    is_fp_zero_point: bool = False

    _cpp_extra_names: ClassVar[tuple[str, ...]] = (
        "is_channel_weight_scale",
        "is_block_weight_scale",
        "is_group_weight_scale",
        "is_tensor_weight_scale",
    )

    def __post_init__(self):
        super().__post_init__()
        if self.weight_scale_type is None:
            if self.weight_scale_group_size_n > 1:
                self.weight_scale_type = WeightScaleType.BLOCK
            elif self.weight_scale_group_size > 0:
                self.weight_scale_type = WeightScaleType.GROUP
            elif self.weight_scale_group_size == 0:
                self.weight_scale_type = WeightScaleType.CHANNEL

        self.is_channel_weight_scale = self.weight_scale_type == WeightScaleType.CHANNEL
        self.is_tensor_weight_scale = self.weight_scale_type in [
            WeightScaleType.TENSOR,
            WeightScaleType.GROUP_TENSOR,
        ]
        self.is_block_weight_scale = self.weight_scale_type == WeightScaleType.BLOCK
        self.is_group_weight_scale = self.weight_scale_type in [
            WeightScaleType.GROUP,
            WeightScaleType.GROUP_TENSOR,
        ]

    @classmethod
    def from_dict(cls, raw_config) -> Self:
        config = super().from_dict(raw_config)
        config.__post_init__()
        return config


@dataclasses.dataclass
class EpilogueConfig(BaseHummingConfig):
    has_bias: bool = False


@dataclasses.dataclass
class MoEConfig(BaseHummingConfig):
    top_k: int = 0
    is_moe: bool = False
    is_moe_down: bool = False

    _name_map = {
        "is_moe": "kIsMoE",
        "is_moe_down": "kIsMoEDown",
    }


@dataclasses.dataclass
class MmaConfig(BaseHummingConfig):
    mma_type: MmaType | None | str = None
    use_f16_accum: bool = False

    @classmethod
    def from_dict(cls, raw_config) -> Self:
        clean_config = cls._preprocess_dict(raw_config)
        config = cls(**clean_config)

        if config.mma_type is None:
            if raw_config["sm_version"] == 90:
                config.mma_type = MmaType.WGMMA
            elif raw_config["sm_version"] >= 75:
                config.mma_type = MmaType.MMA
        elif config.mma_type == "mma":
            config.mma_type = MmaType.MMA
        elif config.mma_type == "wgmma":
            config.mma_type = MmaType.WGMMA

        return config
